src/components/pinnable.js

```diff
@@ -0,0 +1,90 @@
+AFRAME.registerComponent("pinnable", {
+  schema: {
+    pinned: { default: false }
+  },
+
+  init() {
+    this._apply = this._apply.bind(this);
+    this._fireEvents = this._fireEvents.bind(this);
+    this._firePinnedEvents = this._firePinnedEvents.bind(this);
+    this._fireUnpinnedEvents = this._fireUnpinnedEvents.bind(this);
+    this._allowApplyOnceComponentsReady = this._allowApplyOnceComponentsReady.bind(this);
+    this._allowApply = false;
+
+    this.el.sceneEl.addEventListener("stateadded", this._apply);
+    this.el.sceneEl.addEventListener("stateremoved", this._apply);
+
+    // Fire pinned events when we drag and drop or scale in freeze mode,
+    // so transform gets updated.
+    this.el.addEventListener("grab-end", this._firePinnedEvents);
+
+    // Fire pinned events when page changes so we can persist the page.
+    this.el.addEventListener("pager-page-changed", this._firePinnedEvents);
+
+    // Hack: need to wait for the initial grabbable and stretchable components
+    // to show up from the template before applying.
+    this.el.addEventListener("componentinitialized", this._allowApplyOnceComponentsReady);
+    this._allowApplyOnceComponentsReady();
+  },
+
+  remove() {
+    this.el.sceneEl.removeEventListener("stateadded", this._apply);
+    this.el.sceneEl.removeEventListener("stateremoved", this._apply);
+    this.el.removeEventListener("componentinitialized", this._allowApplyOnceComponentsReady);
+  },
+
+  update() {
+    this._apply();
+    this._fireEvents();
+  },
+
+  _fireEvents() {
```

Comment:
Odd to break these into 3 methods. I would just have 1. Also since the events bubble I don't see the need for the extra event on the scene (just check the target of the "pinned" event)

Reply:
the dusty cobwebs of my java-isms using method names to document poking through. i think you're right in this case